controllers/azuremachine_controller.go

```diff
@@ -307,19 +283,26 @@ func (r *AzureMachineReconciler) reconcileNormal(ctx context.Context, machineSco
 		machineScope.Error(errors.New("Failed to create or update VM"), "VM is in failed state", "id", *machineScope.GetVMID())
 		r.Recorder.Eventf(machineScope.AzureMachine, corev1.EventTypeWarning, "FailedVMState", "Azure VM is in failed state")
 		machineScope.SetFailureReason(capierrors.UpdateMachineError)
-		machineScope.SetFailureMessage(errors.Errorf("Azure VM state is %s", vm.State))
+		machineScope.SetFailureMessage(errors.Errorf("Azure VM state is %s", machineScope.GetVMState()))
 		conditions.MarkFalse(machineScope.AzureMachine, infrav1.VMRunningCondition, infrav1.VMProvisionFailedReason, clusterv1.ConditionSeverityWarning, "")
+		// If VM failed provisioning, delete it so it can be recreated
```

Comment:
not sure if we actually want to do this (maybe we should look into Machine Health Checks instead)? but this has been in the reconciler since basically v1alpha1 (moved from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/827/files#diff-fb5deef0b083d17ab538518df2570d4bL266) so I just left it for now to not burry behavior breaking changes in a huge refactor

Reply:
Agreed. Another PR to track the MHC work as well as an e2e test would be super cool